File: src/tree.cpp
#include <algorithm>
#include "learner.h"
#include "tree.h"

#include <iostream>
#include <typeinfo>

using namespace std;

namespace cuncuno {


void clearBindings(ArrivalNode& node, Count reserved);
void clearBindings(StorageNode& node, Count reserved);

void receive(LeafNode& node, Binding& binding);
void receive(PredicateNode& node, Binding& binding);
void receive(RootNode& node, Binding& binding);
void receive(VariableNode& node, Binding& binding);

/**
 * Propagates bindings with direct storage through the tree.
 */
template<typename SomeNode>
void receive(SomeNode& node, std::vector<Binding>& bindings) {
  clearBindings(node, bindings.size());
  for (Count b = 0; b < bindings.size(); b++) {
    receive(node, bindings[b]);
  }
}

/**
 * Propagates bindings given as pointers through the tree.
 */
template<typename SomeNode>
void receive(SomeNode& node, std::vector<Binding*>& bindings) {
  clearBindings(node, bindings.size());
  for (Count b = 0; b < bindings.size(); b++) {
    receive(node, *bindings[b]);
  }
}


/// ArrivalNode.

ArrivalNode::ArrivalNode(): arrival(new BindingArrival) {}

ArrivalNode::ArrivalNode(const ArrivalNode& other):
  Node(other), arrival(other.arrival)
{
  // We are another client for these arrivals.
  arrival->acquire();
}

ArrivalNode::~ArrivalNode() {
  arrival->release();
}

void clearBindings(ArrivalNode& node, Count reserved) {
  node.arrival->bindings.clear();
  node.arrival->bindings.reserve(reserved);
}

void ArrivalNode::propagateTo(Node& node) {
  node.propagate(arrival->bindings);
}


/// Binding.

Binding::Binding() {}

Binding::Binding(const Binding& $previous, const void* entity):
  previous(&$previous), entityOrSample(entity) {}

Binding::Binding(const Sample& sample): previous(0), entityOrSample(&sample) {}

Count Binding::count() const {
  Count count(0);
  for (
    const Binding* binding(this); binding->previous; binding = binding->previous
  ) {
    count++;
  }
  return count;
}

const bool Binding::entities(std::vector<const void*>& buffer) const {
  bool allGood(true);
  Count count = this->count();
  buffer.clear();
  buffer.resize(count);
  for (
    const Binding* binding(this); binding->previous; binding = binding->previous
  ) {
    // Fill in reverse order, so subtract count as we go. Also, no need to check
    // bounds. We already made sure we had the right size.
    buffer[--count] = binding->entityOrSample;
    if (!binding->entityOrSample) {
      allGood = false;
    }
  }
  return allGood;
}

/**
 * The entity at the current level, or 0 if at the sample.
 */
const void* Binding::entity() const {
  return previous ? entityOrSample : 0;
}

const Sample& Binding::sample() const {
  const Binding* binding(this);
  while (binding->previous) {
    binding = binding->previous;
  }
  return *reinterpret_cast<const Sample*>(binding->entityOrSample);
}


/// LeafNode.

LeafNode::LeafNode() {}

LeafNode::LeafNode(const LeafNode& other):
  ArrivalNode(other), probability(other.probability) {}

void LeafNode::accept(NodeVisitor& visitor, void* data) {
  visitor.visit(*this, data);
}

Node* LeafNode::copy() {
  return new LeafNode(*this);
}

void LeafNode::propagate(vector<Binding*>& bindings) {
  receive(*this, bindings);
}

void LeafNode::propagate(vector<Binding>& bindings) {
  receive(*this, bindings);
}

void receive(LeafNode& node, Binding& binding) {
  // No kids for leaves, so just accept the binding.
  node.arrival->bindings.push_back(&binding);
}


/// Node.

Node::Node(Id $id): id($id), $parent(0) {}

Node::Node(const Node& other): id(other.id), $parent(0) {
  for (
    vector<Node*>::const_iterator k(other.kids.begin());
    k != other.kids.end();
    k++
  ) {
    Node* kid = (*k)->copy();
    kid->$parent = this;
    // Don't use pushKid, because we want to keep the ID in this case.
    kids.push_back(kid);
  }
}

Node::~Node() {
  while (!kids.empty()) {
    delete kids.back();
    kids.pop_back();
  }
  // Clear for clean data in case accidentally used.
  id = 0;
  $parent = 0;
}

Node* Node::findById(Node::Id id) {
  if (this->id == id) {
    return this;
  }
  for (vector<Node*>::iterator k = kids.begin(); k != kids.end(); k++) {
    Node* node = (*k)->findById(id);
    if (node) {
      return node;
    }
  }
  // No match found.
  return 0;
}

void Node::insertParent(Node& parent, Count index) {
  // Make sure we have a spot for this kid.
  if (index >= parent.kids.size()) {
    throw "No such kid index.";
  }
  // Put the new parent in the current parent.
  Node* currentParent(this->parent());
  if (currentParent) {
    vector<Node*>& siblings = currentParent->kids;
    vector<Node*>::iterator location =
      find(siblings.begin(), siblings.end(), this);
    if (location == siblings.end()) {
      throw "Node not in parent.";
    }
    parent.id = root()->generateId();
    parent.$parent = currentParent;
    *location = &parent;
  }
  // Purge any kid in the way, then put the new kid there.
  if (parent.kids[index]) {
    // TODO Do I really want joint required here?
    Joint joint;
    parent.kids[index]->purge(joint);
  }
  this->$parent = &parent;
  parent.kids[index] = this;
}

void Node::leaves(std::vector<LeafNode*>& buffer) {
  struct LeafVisitor: NodeVisitor {
    LeafVisitor(std::vector<LeafNode*>& $buffer): buffer($buffer) {}
    virtual void visit(LeafNode& node, void* data) {
      buffer.push_back(&node);
    }
    std::vector<LeafNode*>& buffer;
  };
  LeafVisitor visitor(buffer);
  traverse(visitor);
}

Node* Node::parent() {
  return $parent;
}

void Node::propagate() {
  for (Count k = 0; k < kids.size(); k++) {
    propagateTo(*kids[k]);
  }
}

void Node::purge(Joint& joint) {
  joint.node = this->parent();
  if (joint.node) {
    vector<Node*>& siblings = joint.node->kids;
    // TODO Put this logic in separate method then call it: this->extract();
    // Remove doesn't work for me here. I guess I don't understand it.
    vector<Node*>::iterator self =
      find(siblings.begin(), siblings.end(), this);
    if (self == siblings.end()) {
      throw "Extracted node not in parent.";
    }
    *self = 0;
    joint.location = self;
    siblings.erase(self);
  }
  delete this;
}

void Node::replaceWith(Node& node) {
  // Put the new parent in the current parent.
  Node* currentParent(this->parent());
  if (currentParent) {
    vector<Node*>& siblings = currentParent->kids;
    vector<Node*>::iterator location =
      find(siblings.begin(), siblings.end(), this);
    if (location == siblings.end()) {
      throw "Node not in parent.";
    }
    // Assign ids for the new tree members.
    if (root()) {
      struct IdNodeVisitor: NodeVisitor {
        IdNodeVisitor(RootNode& $root): root($root) {}
        virtual void visit(Node& node, void* data) {
          node.id = root.generateId();
        }
        RootNode& root;
      };
      IdNodeVisitor visitor(*root());
      node.traverse(visitor);
    }
    // Put the node in place.
    node.$parent = currentParent;
    *location = &node;
  }
  // Clean out the old parent link.
  this->$parent = 0;
}

void Node::pushKid(Node& kid) {
  RootNode* root = this->root();
  if (root) {
    kid.id = root->generateId();
  }
  kid.$parent = this;
  kids.push_back(&kid);
}

RootNode* Node::root() {
  Node* highest = this;
  while (highest->parent()) {
    highest = highest->parent();
  }
  return dynamic_cast<RootNode*>(highest);
}

void Node::traverse(NodeVisitor& visitor, void* data) {
  visitor.visit(*this, data);
  accept(visitor, data);
  for (std::vector<Node*>::iterator k = kids.begin(); k != kids.end(); k++) {
    (*k)->traverse(visitor, data);
  }
}


/// PredicateNode.

PredicateNode::PredicateNode() {}

PredicateNode::PredicateNode(const PredicateNode& other):
  // TODO Copy the predicate to make sure the tree really is copied???
  // TODO Could that be slow? Make it instead const and Shared?
  ArrivalNode(other), predicate(other.predicate) {}

void PredicateNode::accept(NodeVisitor& visitor, void* data) {
  visitor.visit(*this, data);
}

Node* PredicateNode::copy() {
  return new PredicateNode(*this);
}

void PredicateNode::propagate(vector<Binding*>& bindings) {
  receive(*this, bindings);
  // TODO Split and propagate.
}

void PredicateNode::propagate(vector<Binding>& bindings) {
  receive(*this, bindings);
  // TODO Split and propagate.
}

void receive(PredicateNode& node, Binding& binding) {
  // Record the binding.
  node.arrival->bindings.push_back(&binding);
}


/// RootNode.

RootNode::RootNode(const RootNode& other):
  StorageNode(other), entityType(other.entityType), nextId(other.nextId) {}

RootNode::RootNode(const Type& $entityType, Node* $kid):
  StorageNode(1), entityType($entityType), kid($kid), nextId(id + 1)
{
  // TODO Delete if we remove vector kids.
  if (kid) {
    pushKid(*kid);
  } else {
    kids.push_back(kid);
  }
}

void RootNode::accept(NodeVisitor& visitor, void* data) {
  visitor.visit(*this, data);
}

void RootNode::bindingsPush(const std::vector<Sample>& samples) {
  storage->bindings.reserve(samples.size());
  for (Count s = 0; s < samples.size(); s++) {
    storage->bindings.push_back(Binding(samples[s]));
  }
}

Node* RootNode::copy() {
  return new RootNode(*this);
}

Node::Id RootNode::generateId() {
  return nextId++;
}

void RootNode::propagate(vector<Binding*>& bindings) {
  receive(*this, bindings);
  propagateTo(*kids.front());
}

void RootNode::propagate(vector<Binding>& bindings) {
  receive(*this, bindings);
  propagateTo(*kids.front());
}

void RootNode::propagate(const std::vector<Sample>& samples) {
  // Reserve but don't clear.
  storage->bindings.clear();
  storage->bindings.reserve(samples.size());
  // Put the samples into bindings and propagate.
  for (
    std::vector<Sample>::const_iterator s = samples.begin();
    s != samples.end();
    s++
  ) {
    Binding binding(*s);
    cuncuno::receive(*this, binding);
  }
  propagateTo(*kids.front());
}

void receive(RootNode& node, Binding& binding) {
  // Just store the binding.
  node.storage->bindings.push_back(binding);
}


/// StorageNode.

StorageNode::StorageNode(Id id): Node(id), storage(new BindingStorage) {}

StorageNode::StorageNode(const StorageNode& other):
  Node(other), storage(other.storage)
{
  // We are another client for this storage.
  storage->acquire();
}

StorageNode::~StorageNode() {
  storage->release();
}

void clearBindings(StorageNode& node, Count reserved) {
  node.storage->bindings.clear();
  node.storage->bindings.reserve(reserved);
}

void StorageNode::propagateTo(Node& node) {
  node.propagate(storage->bindings);
}


/// VariableNode.

VariableNode::VariableNode(Node* $kid): kid($kid) {
  // TODO Remove this if we switch away from vector storage.
  if (kid) {
    pushKid(*kid);
  } else {
    kids.push_back(kid);
  }
}

VariableNode::VariableNode(const VariableNode& other): StorageNode(other) {}

void VariableNode::accept(NodeVisitor& visitor, void* data) {
  visitor.visit(*this, data);
}

Node* VariableNode::copy() {
  return new VariableNode(*this);
}

void VariableNode::propagate(vector<Binding*>& bindings) {
  receive(*this, bindings);
  propagateTo(*kids.front());
}

void VariableNode::propagate(vector<Binding>& bindings) {
  receive(*this, bindings);
  propagateTo(*kids.front());
}

void receive(VariableNode& node, Binding& binding) {
  // See which entities are left in the sample for binding.
  bool anyAvailable(false);
  const vector<const void*>& entities(binding.sample().entities);
  for (Count e = 0; e < entities.size(); e++) {
    const void* entity(entities[e]);
    // See if this entity is still available.
    bool available(true);
    const Binding* current(&binding);
    while (current) {
      if (entity == binding.entity()) {
        available = false;
        break;
      }
      current = current->previous;
    }
    if (available) {
      // It is. Remember we got one and store it.
      anyAvailable = true;
      node.storage->bindings.push_back(Binding(binding, entity));
    }
  }
  // See if we need to bind a dummy.
  if (!anyAvailable) {
    node.storage->bindings.push_back(Binding(binding, 0));
  }
}


}

File: js/tree.js
// Imports.


var $class = require('./etc').$class;
var dump = console.log;


// Exports.


exports.emptyBindings = emptyBindings;
exports.startTree = startTree;


// Classes.


function BindingBag(bag) {
  this.bag = bag;
  this.entityLists = [];
}


function Node() {

  this.bindings = [];
  this.id = 0;
  this.parent = null;
  initKids(this);

} $class(Node, null, function() {

  this.clone = function() {
    // Mostly, we're okay keeping things just as prototypes. A bit risky if we
    // don't track the consequences (such as for bindings lists), but we make
    // a lot of clones, and avoiding copying data seems wise and simple.
    var Clone = function() {};
    Clone.prototype = this;
    var node = new Clone;
    var kids = this.kids();
    if (kids.length) {
      // We do want to copy the kids.
      var clonedKids = [];
      for (var k = 0; k < kids.length; k++) {
        clonedKids.push(kids[k]);
      }
      node.setKids(clonedKids);
    }
    return node;
  };

  this.getNode = function(id) {
    if (this.id == id) {
      return this;
    } else {
      var kids = this.kids();
      for (var k = 0; k < kids.length; k++) {
        var node = kids[k].getNode(id);
        if (node) {
          return node;
        }
      }
      return null;
    }
  };

  this.isVar = function() {
    return false;
  };

  this.leaves = function() {
    var result = [];
    pushLeaves(this);
    return result;
    function pushLeaves(node) {
      var kids = node.kids();
      if (kids.length) {
        for (var k = 0; k < kids.length; k++) {
          pushLeaves(kids[k]);
        }
      } else {
        // This is a leaf.
        result.push(node);
      }
    }
  }

  this.newLeaf = function(kid) {
    return new LeafNode(kid);
  };

  this.newSplit = function(kid) {
    return new SplitNode(kid);
  };

  this.newVar = function(kid) {
    return new VarNode(kid);
  };

  this.replaceWith = function(node) {
    if (this.parent) {
      var kids = this.parent.kids();
      for (var k = 0; k < kids.length; k++) {
        var kid = kids[k];
        if (kid.id == this.id) {
          this.parent.setKid(Number(k), node);
        }
      }
    }
  };

  this.varDepth = function() {
    var depth = 0;
    var node = this;
    while (node) {
      depth += node.isVar();
      node = node.parent;
    }
    return depth;
  };

  this.root = function() {
    return this.parent ? this.parent.root() : this;
  }

});



function LeafNode() {

  Node.call(this);
  this.prob = 0;

} $class(LeafNode, Node, function() {

  this.kids = function() {
    return [];
  };

  this.propagate = function(bindings) {
    if (bindings) {
      // We got new bindings.
      this.bindings = bindings;
    }
  };

});


function RootNode(kid) {

  this.kid = kid || new LeafNode;
  // Set nextId before calling super, so kids will init correctly.
  this.nextId = 1;
  Node.call(this);
  this.bags = [];

} $class(RootNode, Node, function() {

  this.kids = function() {
    return [this.kid];
  };

  this.generateId = function() {
    return this.nextId++;
  };

  this.propagate = function(bindings) {
    if (bindings) {
      // We got new bindings.
      this.bindings = bindings;
    } else {
      // We need to use the ones we already had.
      bindings = this.bindings;
    }
    // dump("RootNode to prop " + bindings.length + " bindings");
    this.kid.propagate(bindings);
  };

  this.setKid = function(k, kid) {
    if (k) throw "bad kid index " + k;
    this.kid = kid;
    initKids(this);
  };

  this.setKids = function(kids) {
    this.kid = kids[0];
    initKids(this);
  };

});


function SplitNode($true, $false, error) {

  this.$true = $true || new LeafNode;
  this.$false = $false || new LeafNode;
  this.error = error || new LeafNode;
  Node.call(this);

} $class(SplitNode, Node, function() {

  this.kids = function() {
    return [this.$true, this.false, this.error];
  };

  this.setKid = function(k, kid) {
    var keys = ['$true', '$false', 'error'];
    this[keys[k]] = kid;
    initKids(this);
  };

  this.setKids = function(kids) {
    this.$true = kids[0];
    this.$false = kids[1];
    this.error = kids[2];
    initKids(this);
  };

});



function VarNode(kid) {

  this.kid = kid || new LeafNode;
  Node.call(this);

} $class(VarNode, Node, function() {

  this.isVar = function() {
    return true;
  };

  this.kids = function() {
    return [this.kid];
  };

  this.propagate = function(bindings) {

    if (bindings) {
      // We got new bindings.
      var outgoings = [];
      var inCount = 0;
      var outCount = 0;
      for (var b = 0; b < bindings.length; b++) {
        var binding = bindings[b];
        var bag = binding.bag;
        var outgoing = new BindingBag(bag);
        for (var L = 0; L < binding.entityLists.length; L++) {
          var entities = binding.entityLists[L];
          inCount++;
          var any = false;
          for (var e = 0; e < bag.entities.length; e++) {
            var entity = bag.entities[e];
            if (entities.indexOf(entity) < 0) {
              // We don't repeat bindings in SMRF.
              bind(entity);
              any = true;
            }
          }
          if (!any) {
            // Dummy entity here for when no new entities were available.
            // We send these down error branches later at split nodes.
            bind(null);
          }
        }
        outgoings.push(outgoing);
      }
      this.bindings = outgoings;
      //dump("Built " + inCount + " bindings into " + outCount);
    }

    // Continue propagation.
    this.kid.propagate(this.bindings);

    // Functions.
    function bind(entity) {
      var outEntities = entities.slice();
      outEntities.push(entity);
      outgoing.entityLists.push(outEntities);
      outCount++;
    }

  };

  this.setKid = function(k, kid) {
    if (k) throw "bad kid index " + k;
    this.kid = kid;
    initKids(this);
  };

  this.setKids = function(kids) {
    this.kid = kids[0];
    initKids(this);
  };

});


// Functions.


function emptyBindings(bags) {
  var bindings = [];
  for (var b = 0; b < bags.length; b++) {
    var binding = new BindingBag(bags[b]);
    // We start with one empty binding per bag.
    binding.entityLists.push([]);
    bindings.push(binding);
  }
  return bindings;
}


function initKids(parent) {
  var kids = parent.kids();
  // Set parent.
  for (var k = 0; k < kids.length; k++) {
    kids[k].parent = parent;
  }
  // Set ids if we're part of a tree already.
  var root = parent.root();
  if (root instanceof RootNode) {
    for (var k = 0; k < kids.length; k++) {
      var kid = kids[k];
      if (!kid.id) {
        kid.id = root.generateId();
        initKids(kid);
      }
    }
  }
}


function startTree() {
  return new RootNode;
}
